src/IO/WriteBufferFromS3.h

```diff
@@ -52,75 +44,121 @@ class WriteBufferFromS3 final : public BufferWithOwnMemory<WriteBuffer>
+    void fillUploadRequest(UploadPartTask & task, PartData && data);
     void processUploadRequest(UploadPartTask & task);
 
     struct PutObjectTask;
-    void fillPutRequest(Aws::S3::Model::PutObjectRequest & req);
+    void fillPutRequest(PutObjectTask & task, PartData && data);
     void processPutRequest(const PutObjectTask & task);
 
-    void waitForReadyBackGroundTasks();
+    void collectReadyBackGroundTasks();
     void waitForAllBackGroundTasks();
     void waitForAllBackGroundTasksUnlocked(std::unique_lock<std::mutex> & bg_tasks_lock);
 
     const String bucket;
     const String key;
     const S3Settings::RequestSettings::PartUploadSettings settings;
+    const WriteSettings write_settings;
     const bool check_objects_after_upload = false;
     const size_t max_unexpected_write_error_retries = 4;
     const std::shared_ptr<const Aws::S3::S3Client> client_ptr;
     const std::optional<std::map<String, String>> object_metadata;
+    Poco::Logger * log = &Poco::Logger::get("WriteBufferFromS3");
 
-    size_t upload_part_size = 0;
-    std::shared_ptr<Aws::StringStream> temporary_buffer; /// Buffer to accumulate data.
-    size_t last_part_size = 0;
-    size_t part_number = 0;
+    BufferAllocationPolicy buffer_allocation_policy;
+
+    /// Following fields are for background uploads in thread pool (if specified).
+    /// We use std::function to avoid dependency of Interpreters
+    const ThreadPoolCallbackRunner<void> schedule;
 
     /// Upload in S3 is made in parts.
     /// We initiate upload, then upload each part and get ETag as a response, and then finalizeImpl() upload with listing all our parts.
     String multipart_upload_id;
     std::vector<String> TSA_GUARDED_BY(bg_tasks_mutex) part_tags;
 
-    bool is_prefinalized = false;
-
-    /// Following fields are for background uploads in thread pool (if specified).
-    /// We use std::function to avoid dependency of Interpreters
-    const ThreadPoolCallbackRunner<void> schedule;
-
     std::unique_ptr<PutObjectTask> put_object_task; /// Does not need protection by mutex because of the logic around is_finished field.
+
     std::list<UploadPartTask> TSA_GUARDED_BY(bg_tasks_mutex) upload_object_tasks;
     int num_added_bg_tasks TSA_GUARDED_BY(bg_tasks_mutex) = 0;
     int num_finished_bg_tasks TSA_GUARDED_BY(bg_tasks_mutex) = 0;
 
     std::mutex bg_tasks_mutex;
     std::condition_variable bg_tasks_condvar;
 
-    Poco::Logger * log = &Poco::Logger::get("WriteBufferFromS3");
+    bool multipart_upload_finished = false;
+    bool multipart_upload_aborted = false;
+
+    /// Track that prefinalize() is called only once
+    bool is_prefinalized = false;
+
+    /// First fully filled buffer has to be delayed
+    /// There are two ways after:
+    /// First is to call prefinalize/finalize, which leads to single part upload
+    /// Second is to write more data, which leads to multi part upload
+    std::deque<PartData> detached_part_data;
```

Comment:
If I understood correctly this PR can help to eliminate extra copying to `std::stringstream`, but it can't help with huge memory consumption because you have to keep the data written to `WriteBufferFromS3` anyway until aws sdk uses it.

Reply:
Yes. It is right. To handle it we need bound unfinished tasks. That problem hasn't been addressed here.

Reply:
What are bound unfinished tasks? For me it doesn't look possible to solve the memory consumption issue in general for big `upload_part_size` without making an intermediate copy to a local disk.

Reply:
Well. the simpliest solution could be as the following:
- make one_buffer, push it to thread pool, make account bytes_in_fly += one_buffer.size()
- make other_buffer, push it to thread pool, make account bytes_in_fly += other_buffer.size()
- make third_buffer, if bytes_in_fly is too big, then hung and wait until some task is finished
- for example one is finished, we do bytes_in_fly -= one_buffer.size(), that unblock third_buffer buffer to proceed further
That I meant by "bound unfinished tasks" by it's sizes.

Reply:
AWS SDK uses seeking on passed `std::stream` so this way you'll never be able to deallocate any buffers.

Reply:
> AWS SDK uses seeking on passed `std::stream` so this way you'll never be able to deallocate any buffers until the whole part (which is up to 5GB) is uploaded

Yes. that is correct.

Reply:
> What are the bound unfinished tasks? I doubt that it's possible to solve the memory consumption issue in general for big `upload_part_size` at all without making an intermediate copy to a local disk.

Interesting approach. That will work for sure. In such way we definetly should care about an extra memcpu :)

Reply:
If you manage to get rid of seeking inside AWS SDK in order to calculate checksums, then it would be better to change AWS SDK a bit to teach it work directly with our WriteBuffer, without any intermediate buffering at all. I mean in this case `WriteBufferFromS3` could directly send ready blocks of data to network.